tempodb/tempodb.go

```diff
@@ -309,77 +319,102 @@ func (rw *readerWriter) pollBlocklist() {
+
+	for _, blockID := range blockIDs {
+		bg.Add(1)
+		go func(b uuid.UUID) {
+			defer bg.Done()
+			m, cm, e := rw.pollBlock(ctx, tenantID, b)
+			if e != nil {
+				chErr <- e
+			} else if m != nil {
+				chMeta <- m
+			} else if cm != nil {
+				chCompactedMeta <- cm
 			}
+		}(blockID)
+	}
 
-			// todo:  make this not terrible. this mutex is dumb we should be returning results with a channel. shoehorning this into the worker pool is silly.
-			//        make the worker pool more generic? and reusable in this case
-			listMutex.Lock()
-			if blockMeta != nil {
-				blocklist = append(blocklist, blockMeta)
+	bg.Wait()
+	close(chErr)
+	close(chMeta)
+	close(chCompactedMeta)
 
-			} else if compactedBlockMeta != nil {
-				compactedBlocklist = append(compactedBlocklist, compactedBlockMeta)
-			}
-			listMutex.Unlock()
+	for err := range chErr {
+		return nil, nil, err
+	}
 
-			return nil, nil
-		})
+	newBlockList := make([]*backend.BlockMeta, 0, len(blockIDs))
+	for m := range chMeta {
+		newBlockList = append(newBlockList, m)
+	}
+	sort.Slice(newBlockList, func(i, j int) bool {
+		return newBlockList[i].StartTime.Before(newBlockList[j].StartTime)
+	})
 
-		if err != nil {
-			metricBlocklistErrors.WithLabelValues(tenantID).Inc()
-			level.Error(rw.logger).Log("msg", "run blocklist jobs", "tenantID", tenantID, "err", err)
-			continue
-		}
+	newCompactedBlocklist := make([]*backend.CompactedBlockMeta, 0, len(blockIDs))
+	for cm := range chCompactedMeta {
+		newCompactedBlocklist = append(newCompactedBlocklist, cm)
+	}
+	sort.Slice(newCompactedBlocklist, func(i, j int) bool {
+		return newCompactedBlocklist[i].StartTime.Before(newCompactedBlocklist[j].StartTime)
+	})
 
-		metricBlocklistLength.WithLabelValues(tenantID).Set(float64(len(blocklist)))
+	return newBlockList, newCompactedBlocklist, nil
+}
 
-		sort.Slice(blocklist, func(i, j int) bool {
-			return blocklist[i].StartTime.Before(blocklist[j].StartTime)
-		})
-		sort.Slice(compactedBlocklist, func(i, j int) bool {
-			return compactedBlocklist[i].StartTime.Before(compactedBlocklist[j].StartTime)
-		})
+func (rw *readerWriter) pollBlock(ctx context.Context, tenantID string, blockID uuid.UUID) (*backend.BlockMeta, *backend.CompactedBlockMeta, error) {
+	var compactedBlockMeta *backend.CompactedBlockMeta
+	blockMeta, err := rw.r.BlockMeta(ctx, blockID, tenantID)
+	// if the normal meta doesn't exist maybe it's compacted.
+	if err == backend.ErrMetaDoesNotExist {
+		blockMeta = nil
+		compactedBlockMeta, err = rw.c.CompactedBlockMeta(blockID, tenantID)
+	}
 
-		rw.blockListsMtx.Lock()
-		rw.blockLists[tenantID] = blocklist
-		rw.compactedBlockLists[tenantID] = compactedBlocklist
-		rw.blockListsMtx.Unlock()
+	// blocks in intermediate states may not have a compacted or normal block meta.
+	//   this is not necessarily an error, just bail out
+	if err == backend.ErrMetaDoesNotExist {
+		return nil, nil, nil
 	}
+
+	if err != nil {
+		metricBlocklistErrors.WithLabelValues(tenantID).Inc()
```

Comment:
this error will be double metric'ed.  once here and once in the main poll loop.

Reply:
This is fixed.  This was a misread of the error returned by pool.RunJobs, `"run blocklist jobs"`.  The bounded wait group  doesn't have that class of error (an error in starting the jobs), so this error handling is eliminated entirely.